networkit/cpp/io/KONECTGraphReader.cpp

```diff
@@ -1,226 +1,279 @@
+			if (graphType == "weighted" || graphType == "posweighted" || graphType == "signed"){
+				weighted = true;
+				valuesPerLine = 3;
+			} else if (graphType == "positive"){ //multiple edges
+				if(multipleEdgesHandlingMethod == SUM_WEIGHTS_UP){
+					weighted = true;
+				}
+				multiple = true; //weights will be added
+			} else if (graphType == "multisigned" || graphType == "multiweighted" || graphType == "multiposweighted"){
+				weighted = true;
+				multiple = true;
+				valuesPerLine = 3;
+			} else if (graphType == "dynamic"){
+				throw std::runtime_error("Dynamic graphs are not supported yet");
+			} else if (graphType != "unweighted"){
+				throw std::runtime_error("Graph weight not tagged properly. Weight is: " + graphType);
+			}
+			DEBUG("First property line read in. Format: "+graphFormat+" / Type: "+graphType);
+			if (graphFormat == "bip"){
+				INFO("Your graph is flagged as a bipartite one. Keep in mind that"
+				 "NetworKit cannot handle this kind of graph in its special cases."
+				 "It is imported as a usual undirected graph and edges might be discarded.");
+			}
+			if(multiple){
+				DEBUG("Selected handling method for multiple edges is: "+std::to_string(multipleEdgesHandlingMethod));
+			}
+		} else {
+			throw std::runtime_error("No graph properties line found. This line is mandatory.");
+		}
+
+		skipWhitespace();
+		if (*it != '\n'){
+			throw std::runtime_error("No break symbol after first property line");
+		}else{
+			++it;
+		}
+		skipWhitespace();
+		//second optional property line
+		if(*it == '%'){
+			++it;
+			skipWhitespace();
+			numberOfEdges = scanId();
+			skipWhitespace();
+			numberOfNodes = scanId();
+			secondPropertyLine = true;
+			while(it != end && *it != '\n')
+				++it;
+			if(it >= end){ // proper error message if file ends unexpected
+				throw std::runtime_error("Unexpected end of file");
+			}
+			if (*it != '\n'){
+				throw std::runtime_error("File not properly formatted. Last character in second property line is: "+std::string(1, *it));
+			}else{
+				++it;
+			}
+			DEBUG("Second property line read in. Edges: "+std::to_string(numberOfEdges)+ " / Nodes: "+std::to_string(numberOfNodes));
+		}
+
+
+		Graph graph((secondPropertyLine ? numberOfNodes : 0), weighted, directed);
+
+		//Map nodes and increase graph size if no second property is defined
+		std::function<node(node)> mapNode = [&] (node in) -> node {
+			if(secondPropertyLine){
+				if (in > numberOfNodes){ //if file is corrupted
+					//secondPropertyLine is made useless
+					ERROR("Given number of nodes by file does not match actual graph");
+					secondPropertyLine = false;
+					if(remapNodes){ // if remapNodes is selected true, the map has to be initalized with the existing nodes
+						nodeIdMap.reserve(numberOfNodes);
+						graph.forNodes([&](node u) {
+							nodeIdMap.insert({u,u});
+						});
+					}
+					return mapNode(in);
+				}else{
```

Comment:
Shouldn't there `remapNodes` also be considered when the number of nodes is given in the graph file? I.e., shouldn't this be the other way around, if `remapNodes` is set always use `nodeIdMap` and only otherwise possibly add nodes to the graph? At least this is what I would expect from the documentation of the constructor.

Reply:
Well, the whole use of the remapNodes parameter is to allow the user to read in KONECT graphs with non-consecutive ID's. If the number of nodes is given in the second property line, there is basically no need to remap Nodes (because we can assume that the id's are consecutive with a VERY high probability), which would cost more runtime and memory to additionally map nodes in a map.

The reason I added the remap behaviour in this error handling is, that if the second property line is "lying" about the actual number of nodes, the user can still remap the nodes and thereby decrease the amount if isolated nodes and size of the graph, without greatly modifying the graph file itself.